Add sensor for AWS Batch (#19850)

yeshbash opened the pull request:
Adds a sensor implementation to ask for the status of an
AWS Batch job. The sensor will enable DAGs to wait for the
batch job to reach a terminal state before proceeding to the
downstream tasks.

closes: #19850

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

Commit:
Add sensor for AWS Batch (#19850)

Adds a sensor implementation to ask for the status of an
AWS Batch job. The sensor will enable DAGs to wait for the
batch job to reach a terminal state before proceeding to the
downstream tasks.

eladkal commented on airflow/providers/amazon/aws/sensors/batch.py:
```diff
@@ -0,0 +1,85 @@
+        'RUNNING',
+    )
+    FAILURE_STATES = ('FAILED',)
+    SUCCESS_STATES = ('SUCCEEDED',)
```
I think it would be better to manage the statuses on the hook level.
The hook itself also has a function that require the same definitions:
https://github.com/apache/airflow/blob/0df50f42dde4bd9b4c99cb6646416dde6fd4961e/airflow/providers/amazon/aws/hooks/batch_client.py#L248-L255

So we should have one place to define it for both.

yeshbash replied:
Good point - I will define these in `AwsBatchClientHook` and reference it.

While making the change, I noticed that the hook also uses two additional combinations of state values for its own methods

1. `poll_for_job_running`: checks for `["RUNNING", "SUCCEEDED", "FAILED"]`.
2. `poll_for_job_complete` : checks for `["SUCCEEDED", "FAILED"]`

I'm planning to create the following state definitions to cater to all cases. Let me know if you think otherwise and leaving the above two methods the same would be better.
```
    FAILURE_STATE = 'FAILED'
    SUCCESS_STATE = 'SUCCEEDED'
    RUNNING_STATE = 'RUNNING'
    INTERMEDIATE_STATES = (
        'SUBMITTED',
        'PENDING',
        'RUNNABLE',
        'STARTING',
        RUNNING_STATE,
    )
```

ferruzzi commented on airflow/providers/amazon/aws/sensors/batch.py:
```diff
@@ -0,0 +1,85 @@
+            raise AirflowException(f'Batch sensor failed. Batch Job Status: {state}')
+
+        if state in self.INTERMEDIATE_STATES:
+            return False
```
I'm not overly familiar with Batch yet.  Is there a use case for setting a target state?  Rather than just returning success/working/failed, is it desirable to have the option to check if it is currently RUNNING, for example?

yeshbash replied:
This sensor's objective is to poke until the job reaches a logical end - success/failure. In this context, I can't think of a use-case where matching against any other target state can be useful. There could be other places where we want to check against a target state and the AwsBatchClientHook already has a method for this.

https://github.com/apache/airflow/blob/0df50f42dde4bd9b4c99cb6646416dde6fd4961e/airflow/providers/amazon/aws/hooks/batch_client.py#L321

Let me know if I misunderstood your comment.

yeshbash replied:
I was looking at emr_sensor and that helped me understand your question better. Are you asking if making the terminal states configurable such that the sensor can return true/false based on it?

I don't know a definitive answer but this is my thought process
- I could see it useful for DAGs which follow a fire-and-forget pattern to launch the job.
- In fire-and-forget scenarios, would there be a need to use a Sensor? - I don't think so
- In the case of EMR, a cluster RUNNING could be a logical terminal state for long-running clusters. On the contrary, the batch is for ephemeral tasks and the same might not make sense
- I see `AwsBatchOperator` providing this option through the `waiter` argument.

With this, I'm inclining to say that a configurable terminal state might not be useful here. Let me know your thoughts

ferruzzi replied:
I appreciate the detailed reply; that makes sense.

As an example of what I was thinking: the EKS Nodegroup sensor can be used to wait for the nodegroup to go "creating" so it can send the command to start spinning up a second, or wait to go "active" and trigger a task to run on it, or wait to go "nonexistent" so the cluster it is attached to could be deleted.   I was wondering if Batch had a usecase for adding similar logic, but it sounds like that may not be the case.  

Carry on, and sorry for the distraction. 😄

Commit:
Define job status inside AwsBatchClientHook

Different Aws Batch job statuses are used in both AwsBatchClientHook and
BatchSensor. This change defines the status strings inside
AwsBatchClientHook such that it can be reused across other classes

eladkal reviewed (approved):
LGTM
@yeshbash would you mind adding also example dag and docs?
you can use [GCS](https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-google/operators/cloud/gcs.rst#sensors) docs as reference
